dash-renderer/src/observers/requestedCallbacks.ts

```diff
@@ -0,0 +1,263 @@
+            duplicates will be removed for real
+        */
+        requested = difference(requested, rDuplicates);
+
+        /*
+            2. Remove duplicated `prioritized`, `executing` and `watching` callbacks
+        */
+
+        /*
+            Extract all but the first callback from each IOS-key group
+            these callbacks are `prioritized` and duplicates.
+        */
+        const pDuplicates = flatten(map(
+            group => group.slice(0, -1),
+            values(
+                groupBy<ICallback>(
+                    getUniqueIdentifier,
+                    concat(prioritized, requested)
+                )
+            )
+        ));
+
+        const eDuplicates = flatten(map(
+            group => group.slice(0, -1),
+            values(
+                groupBy<ICallback>(
+                    getUniqueIdentifier,
+                    concat(executing, requested)
+                )
+            )
+        )) as IExecutingCallback[];
+
+        const wDuplicates = flatten(map(
+            group => group.slice(0, -1),
+            values(
+                groupBy<ICallback>(
+                    getUniqueIdentifier,
+                    concat(watched, requested)
+                )
+            )
+        )) as IExecutingCallback[];
+
+        /*
+            3. Modify or remove callbacks that are outputing to non-existing layout `id`.
+        */
+
+        const { added: rAdded, removed: rRemoved } = pruneCallbacks(requested, paths);
+        const { added: pAdded, removed: pRemoved } = pruneCallbacks(prioritized, paths);
+        const { added: eAdded, removed: eRemoved } = pruneCallbacks(executing, paths);
+        const { added: wAdded, removed: wRemoved } = pruneCallbacks(watched, paths);
+
+        /*
+            TODO?
+            Clean up the `requested` list - during the dispatch phase,
+            it will be updated for real
+        */
+        requested = concat(
+            difference(
+                requested,
+                rRemoved
+            ),
+            rAdded
+        );
+
+        /*
+            4. Find `requested` callbacks that do not depend on a outstanding output (as either input or state)
+        */
+        let readyCallbacks = getReadyCallbacks(requested, pendingCallbacks);
+
+        /*
+            If:
+            - there are `requested` callbacks
+            - no `requested` callback can be promoted to `prioritized`
+            - no callbacks are `prioritized`, `executing`, `watched` or `executed`
+            Then:
+            - the `requested` callbacks form a ciruclar dependency and can never be executed
+            - prune them out of `requested`
+        */
+        const rCircular = (
+            !readyCallbacks.length &&
+            !prioritized.length &&
+            !executing.length &&
+            !watched.length &&
+            !executed.length &&
+            requested.length
+        ) ? requested : [];
```

Comment:
This handling of circular dependencies may need to change as it is different from the current undocumented behavior. I would suggest a potential approach of pruning out callback definitions in a circular dependency from the graph entirely, allowing for no ambiguity.